### DISCUSSION TITLE
Testing methodology

### DISCUSSION THREAD

**Discussion by mgrider** [kean/Get#71] [General]
In my application, I have a single class `DefaultNetworkService` that handles all the network requests. (About the name: `DefaultNetworkService` conforms to a `NetworkService` protocol, so it can be mocked when testing other parts of the app.)

In general, I've tried to leave most of the app's logic out of `DefaultNetworkService`, but inevitably some has crept into it, and it would be nice to write tests for it. 

Worth noting it has a `client: Get.APIClient` property. My initial thought was to write a protocol that `Get.APIClient` already conforms to, and inject a mock client. The mock client was feeling like it was going to be a lot of work (even though it really only needs the `send` function).

Instead I took inspiration from the tests in the `Get -> Tests -> GetTests` folder, specifically the `APIClient` extension in `Helpers.swift`.

I quickly made my own `APIClient` extension, but then realized that it's this line that was doing the magic there:

```Swift
            $0.sessionConfiguration.protocolClasses = [MockingURLProtocol.self]
```

I was tempted to include the whole `Mocker` folder in my project, but I couldn't find a way to do that without copying it in. (Would be curious to hear if anyone knows if that is possible.) Instead I decided I could get away with just a customized version of `MockingURLProtocol`. Mine looks like this:

```Swift
open class MockURLProtocol: URLProtocol {

    public var cacheStoragePolicy: URLCache.StoragePolicy = .notAllowed

    static var requestError: Error?
    static var responseValue: Data?
    static var responseStatus = 200
    static var responseHeaders = [String: String]()

    static var requestedUrlStrings = [String]()

    func mockResponse() -> HTTPURLResponse {
        let response = HTTPURLResponse(
            url: URL(string: "https://example.com")!,
            statusCode: MockURLProtocol.responseStatus,
            httpVersion: "HTTP/2.0",
            headerFields: MockURLProtocol.responseHeaders)!
        return response
    }

    private var responseWorkItem: DispatchWorkItem?

    /// Returns Mocked data based on the mocks register in the `Mocker`. Will end up in an error when no Mock data is found for the request.
    override public func startLoading() {
        if let data = MockURLProtocol.responseValue {
            self.client?.urlProtocol(self, didReceive: mockResponse(), cacheStoragePolicy: cacheStoragePolicy)
            self.client?.urlProtocol(self, didLoad: data)
            self.client?.urlProtocolDidFinishLoading(self)
        } else if let error = MockURLProtocol.requestError {
            self.client?.urlProtocol(self, didFailWithError: error)
        } else {
            self.client?.urlProtocol(self, didReceive: mockResponse(), cacheStoragePolicy: cacheStoragePolicy)
            self.client?.urlProtocol(self, didLoad: MockAPIResponseEmpty.data())
            self.client?.urlProtocolDidFinishLoading(self)
        }
    }

    /// Implementation does nothing, but is needed for a valid inheritance of URLProtocol.
    override public func stopLoading() {
        responseWorkItem?.cancel()
    }

    /// Simply sends back the passed request. Implementation is needed for a valid inheritance of URLProtocol.
    override public class func canonicalRequest(for request: URLRequest) -> URLRequest {
        return request
    }

    /// Overrides needed to define a valid inheritance of URLProtocol. We store the url string here.
    override public class func canInit(with request: URLRequest) -> Bool {
        if let urlString = request.url?.path {
            MockURLProtocol.requestedUrlStrings.append(urlString)
        }
        return true
}
```

This is working great, _except_ that, because it's relying on static properties, my tests can't run in parallel. I tried to think of a way to key the canned `responseValue` and `requestError`s, but just didn't find a good way to do it. Particularly if (as I do), I want to check for `requestedUrlStrings` after a request. (Essentially, some of my API requests kick off other requests, and this is some of the logic I'm testing for.) 

For now, I've just decided that all my network tests will happen in one (rather large) test case. It's not great, but it's way better than no tests at all.

Curious if anyone has any feedback on this approach, or thoughts on how to avoid the monolithic test case situation. I'm happy to post more examples if anyone is interested in seeing more of the code.

**Comment by kean** [comment]
Hey,

> I was tempted to include the whole Mocker folder in my project

You can install it as a Swift package from [WeTransfer/Mocker](https://github.com/WeTransfer/Mocker). It's embedded in the project to make the dependency private.

I briefly looked into the `Mocker` code, and it appears to be thread-safe. Now, I'm not sure what's the best way to isolate the tests. The mocks are registered for specific URLs, so that's one potential avenue. I would suggest trying to use different domains for each unit test.

**Reply by mgrider** [reply]
YES! The URL idea is a good one! I will try it on Monday.

**Reply by mgrider** [reply]
And I do also appreciate the pointer to the official Mocker repo. I do generally try and keep dependencies to a minimum, but since this is just for the test target, maybe another is not so bad.
